test/test_domains.py

```diff
@@ -21,4 +21,37 @@ def test_pickle(expr):
     pickle.dump(x, f)
     f.seek(0)
     y = pickle.load(f)
-    assert y == x  # TODO promote to "y is x"
+    assert y is x
+
+
+def test_cache():
+    assert Bint[1] is Bint[1]
+    assert Real is Reals[()]
+    assert Reals[2, 3] is Reals[2, 3]
+
+
+def test_subclass():
+    assert issubclass(Bint, Bint)
+    assert issubclass(Bint[1], Bint)
+    assert issubclass(Bint[1], Bint[1])
+    assert issubclass(Bint[2], Bint)
+    assert issubclass(Bint[2], Bint[2])
+    assert not issubclass(Bint, Bint[1])
+    assert not issubclass(Bint, Bint[2])
+    assert not issubclass(Bint[1], Bint[2])
+    assert not issubclass(Bint[2], Bint[1])
+
+    assert issubclass(Reals, Reals)
+    assert issubclass(Real, Real)
+    assert issubclass(Real, Reals)
+    assert issubclass(Reals[2], Reals)
+    assert issubclass(Reals[2], Reals[2])
+    assert not issubclass(Reals, Real)
+    assert not issubclass(Reals, Real[2])
+    assert not issubclass(Real, Reals[2])
+    assert not issubclass(Real[2], Real)
```

Comment:
Should `Real[2]` error in favor of `Reals[2]`?

Reply:
Yes, indeed I've already included that fix in #354. Do you mind if we handle that in the follow-up PR?